validator/coordinator/src/test/java/tech/pegasys/teku/validator/coordinator/performance/PerformanceTrackerTest.java

```diff
@@ -0,0 +1,258 @@
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.storage.client.ChainUpdater;
+import tech.pegasys.teku.storage.client.MemoryOnlyRecentChainData;
+import tech.pegasys.teku.storage.client.RecentChainData;
+import tech.pegasys.teku.util.config.Constants;
+
+public class PerformanceTrackerTest {
+
+  private final DataStructureUtil dataStructureUtil = new DataStructureUtil();
+  private final StatusLogger log = mock(StatusLogger.class);
+  private final RecentChainData recentChainData =
+      MemoryOnlyRecentChainData.create(mock(EventBus.class));
+  private RecentChainDataPerformanceTracker performanceTracker =
+      new RecentChainDataPerformanceTracker(recentChainData, log);
+  private static final List<BLSKeyPair> VALIDATOR_KEYS = BLSKeyGenerator.generateKeyPairs(64);
+  private final ChainBuilder chainBuilder = ChainBuilder.create(VALIDATOR_KEYS);
+  private final ChainUpdater chainUpdater = new ChainUpdater(recentChainData, chainBuilder);
+
+  @BeforeAll
+  static void setUp() {
+    Constants.SLOTS_PER_EPOCH = 4;
+  }
+
+  @BeforeEach
+  void beforeEach() {
+    chainUpdater.initializeGenesis();
+    performanceTracker.start(UInt64.ZERO);
+  }
+
+  @Test
+  void shouldDisplayPerfectBlockInclusion() {
+    chainUpdater.updateBestBlock(chainUpdater.advanceChainUntil(10));
+    performanceTracker.saveSentBlock(chainUpdater.chainBuilder.getBlockAtSlot(1));
+    performanceTracker.saveSentBlock(chainUpdater.chainBuilder.getBlockAtSlot(2));
+    performanceTracker.onSlot(compute_start_slot_at_epoch(BLOCK_PERFORMANCE_EVALUATION_INTERVAL));
+    BlockPerformance expectedBlockPerformance = new BlockPerformance(2, 2);
+    verify(log).performance(expectedBlockPerformance.toString());
+  }
+
+  @Test
+  void shouldDisplayOneMissedBlock() {
+    chainUpdater.updateBestBlock(chainUpdater.advanceChainUntil(10));
+    performanceTracker.saveSentBlock(chainUpdater.chainBuilder.getBlockAtSlot(1));
+    performanceTracker.saveSentBlock(chainUpdater.chainBuilder.getBlockAtSlot(2));
+    performanceTracker.saveSentBlock(dataStructureUtil.randomSignedBeaconBlock(3));
+    performanceTracker.onSlot(compute_start_slot_at_epoch(BLOCK_PERFORMANCE_EVALUATION_INTERVAL));
+    BlockPerformance expectedBlockPerformance = new BlockPerformance(2, 3);
+    verify(log).performance(expectedBlockPerformance.toString());
+  }
+
+  @Test
+  void shouldDisplayPerfectAttestationInclusion() {
+    chainUpdater.updateBestBlock(chainUpdater.advanceChainUntil(1));
+
+    ChainBuilder.BlockOptions block1Options = ChainBuilder.BlockOptions.create();
+    Attestation attestation1 =
+        chainBuilder
+            .streamValidAttestationsForBlockAtSlot(2)
+            .filter(
+                a ->
+                    a.getData()
+                        .getBeacon_block_root()
+                        .equals(chainBuilder.getBlockAtSlot(1).getRoot()))
+            .findFirst()
+            .get();
```

Comment:
looks like a function with taking 2 integers to get the attestation you need. might be worth a private function to clean up this portion so you don't get lost in setup...

Reply:
Definetely. Done.